src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,232 @@
+#include "Fingerprint.hpp"
+#include "../helpers/Log.hpp"
+#include "../config/ConfigManager.hpp"
+
+#include <filesystem>
+#include <unistd.h>
+#include <pwd.h>
+
+#include <cstring>
+#include <thread>
+
+static const auto FPRINT        = "net.reactivated.Fprint";
+static const auto DEVICE        = "net.reactivated.Fprint.Device";
+static const auto MANAGER       = "net.reactivated.Fprint.Manager";
+static const auto LOGIN_MANAGER = "org.freedesktop.login1.Manager";
+static const auto RETRY_MESSAGE = "Could not match fingerprint. Try again.";
+
+CFingerprint::CFingerprint() {
+    static auto* const PFINGERPRINTREADY   = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:fingerprint_ready_message"));
+    m_sFingerprintReady                    = *PFINGERPRINTREADY;
+    static auto* const PFINGERPRINTPRESENT = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:fingerprint_present_message"));
+    m_sFingerprintPresent                  = *PFINGERPRINTPRESENT;
+    static auto* const PENABLEFINGERPRINT  = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:enable_fingerprint");
+    m_bEnabled                             = **PENABLEFINGERPRINT;
+}
+
+static void fingerprintCheckCallback(std::shared_ptr<CTimer> self, void* data) {
+    if (g_pFingerprint->isAuthenticated())
+        g_pHyprlock->unlock();
+}
+
+void CFingerprint::start() {
+    if (!m_bEnabled)
+        return;
+    std::thread([this]() {
+        // For grace or SIGUSR1 unlocks
+        if (g_pHyprlock->isUnlocked())
+            return;
```

Comment:
That does not hurt but I think it is also not needed.

Reply:
Removed

Reply:
> That does not hurt
> **I think** it is also not needed

Unless you **know**, I consider this a reason for keeping those lines in.

Reply:
well i know that **should not** be needed, because a race, where we go through that check and then hyprlock is unlocked, (via grace or USR1) might cause some problems. So even if the check fixes smth, it does not fix it properly.

Reply:
@moggiesir is it fine when the main thread calls `terminate` before `m_sDBUSState.connection = sdbus::createSystemBusConnection();` succeeds? Also is `leaveEventLoop` thread safe?

Reply:
It's fine when the main thread calls `terminate` before `m_sDBUSState.connection = sdbus::createSystemBusConnection();` succeeds since `terminate` checks if the pointer is initialized. I also made sure it's fine to call `leaveEventLoop` even if `enterEventLoop` was never called.

For `leaveEventLoop` thread safety, from the [docs](https://github.com/Kistler-Group/sdbus-cpp/blob/master/docs/using-sdbus-c++.md):

> An event loop that blocks in the synchronous enterEventLoop() call can be unblocked through leaveEventLoop() call on the respective bus connection issued from a different thread or from an OS signal handler.

Further, I changed to return the thread from `CFingerprint::start()` so that hyprlock will wait for the thread to terminate to make sure we don't try to destroy the dbus connection while some call is being made.

Reply:
Nice, thanks!